File: scripts/flutter/template/flutter_bloc_template.py
#!/usr/bin/env python3
"""Flutter BLoC 템플릿 생성 스크립트"""

import sys


def to_snake_case(name: str) -> str:
    """PascalCase를 snake_case로 변환"""
    result = []
    for i, char in enumerate(name):
        if char.isupper() and i > 0:
            result.append("_")
        result.append(char.lower())
    return "".join(result)


def generate_bloc(bloc_name: str) -> str:
    """Bloc 메인 파일 템플릿 코드 생성"""
    snake_name = to_snake_case(bloc_name)

    return f"""import 'package:flutter_leaf_kit/flutter_leaf_kit_common.dart';
import 'package:flutter_leaf_kit/flutter_leaf_kit_state.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part '{snake_name}_bloc.freezed.dart';
part '{snake_name}_event.dart';
part '{snake_name}_state.dart';

class {bloc_name}Bloc extends Bloc<{bloc_name}Event, {bloc_name}State> {{
  {bloc_name}Bloc() : super(const {bloc_name}Initial()) {{
    on<{bloc_name}LoadRequested>(_onLoadRequested);
    on<{bloc_name}RefreshRequested>(_onRefreshRequested);
  }}

  Future<void> _onLoadRequested(
    {bloc_name}LoadRequested event,
    Emitter<{bloc_name}State> emit,
  ) async {{
    emit(const {bloc_name}Loading());
    try {{
      // TODO: 데이터 로드 로직 구현
      emit(const {bloc_name}Loaded());
    }} catch (e) {{
      emit({bloc_name}Error(message: e.toString()));
    }}
  }}

  Future<void> _onRefreshRequested(
    {bloc_name}RefreshRequested event,
    Emitter<{bloc_name}State> emit,
  ) async {{
    emit(const {bloc_name}Loading());
    try {{
      // TODO: 데이터 새로고침 로직 구현
      emit(const {bloc_name}Loaded());
    }} catch (e) {{
      emit({bloc_name}Error(message: e.toString()));
    }}
  }}
}}
"""


def generate_event(bloc_name: str) -> str:
    """Event 파일 템플릿 코드 생성"""
    snake_name = to_snake_case(bloc_name)

    return f"""part of '{snake_name}_bloc.dart';

@freezed
sealed class {bloc_name}Event with _${bloc_name}Event {{
  const factory {bloc_name}Event.loadRequested() = {bloc_name}LoadRequested;
  const factory {bloc_name}Event.refreshRequested() = {bloc_name}RefreshRequested;
}}
"""


def generate_state(bloc_name: str) -> str:
    """State 파일 템플릿 코드 생성"""
    snake_name = to_snake_case(bloc_name)

    return f"""part of '{snake_name}_bloc.dart';

@freezed
sealed class {bloc_name}State with _${bloc_name}State {{
  const factory {bloc_name}State.initial() = {bloc_name}Initial;
  const factory {bloc_name}State.loading() = {bloc_name}Loading;
  const factory {bloc_name}State.loaded() = {bloc_name}Loaded;
  const factory {bloc_name}State.error({{required String message}}) = {bloc_name}Error;
}}
"""


def main():
    if len(sys.argv) < 2:
        print("Usage: bloc_template.py <BlocName> [bloc|event|state]", file=sys.stderr)
        print("Example: bloc_template.py Home", file=sys.stderr)
        print("Example: bloc_template.py Home bloc", file=sys.stderr)
        print("Example: bloc_template.py Home event", file=sys.stderr)
        print("Example: bloc_template.py Home state", file=sys.stderr)
        sys.exit(1)

    bloc_name = sys.argv[1]

    # Bloc 접미사가 있으면 제거
    if bloc_name.endswith("Bloc"):
        bloc_name = bloc_name[:-4]

    file_type = sys.argv[2] if len(sys.argv) > 2 else "bloc"

    if file_type == "bloc":
        print(generate_bloc(bloc_name))
    elif file_type == "event":
        print(generate_event(bloc_name))
    elif file_type == "state":
        print(generate_state(bloc_name))
    else:
        print(f"Unknown file type: {file_type}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()

File: scripts/flutter/template/flutter-bloc-template.mjs
#!/usr/bin/env node
// Flutter BLoC 템플릿 생성 스크립트

import process from 'node:process';

function toSnakeCase(name) {
  let out = '';
  for (let i = 0; i < name.length; i++) {
    const ch = name[i];
    if (ch >= 'A' && ch <= 'Z' && i > 0) {
      out += '_';
    }
    out += ch.toLowerCase();
  }
  return out;
}

function generateBloc(blocName) {
  const snakeName = toSnakeCase(blocName);

  return `import 'package:flutter_leaf_kit/flutter_leaf_kit_common.dart';
import 'package:flutter_leaf_kit/flutter_leaf_kit_state.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

part '${snakeName}_bloc.freezed.dart';
part '${snakeName}_event.dart';
part '${snakeName}_state.dart';

class ${blocName}Bloc extends Bloc<${blocName}Event, ${blocName}State> {
  ${blocName}Bloc() : super(const ${blocName}Initial()) {
    on<${blocName}LoadRequested>(_onLoadRequested);
    on<${blocName}RefreshRequested>(_onRefreshRequested);
  }

  Future<void> _onLoadRequested(
    ${blocName}LoadRequested event,
    Emitter<${blocName}State> emit,
  ) async {
    emit(const ${blocName}Loading());
    try {
      // TODO: 데이터 로드 로직 구현
      emit(const ${blocName}Loaded());
    } catch (e) {
      emit(${blocName}Error(message: e.toString()));
    }
  }

  Future<void> _onRefreshRequested(
    ${blocName}RefreshRequested event,
    Emitter<${blocName}State> emit,
  ) async {
    emit(const ${blocName}Loading());
    try {
      // TODO: 데이터 새로고침 로직 구현
      emit(const ${blocName}Loaded());
    } catch (e) {
      emit(${blocName}Error(message: e.toString()));
    }
  }
}
`;
}

function generateEvent(blocName) {
  const snakeName = toSnakeCase(blocName);

  return `part of '${snakeName}_bloc.dart';

@freezed
sealed class ${blocName}Event with _$${blocName}Event {
  const factory ${blocName}Event.loadRequested() = ${blocName}LoadRequested;
  const factory ${blocName}Event.refreshRequested() = ${blocName}RefreshRequested;
}
`;
}

function generateState(blocName) {
  const snakeName = toSnakeCase(blocName);

  return `part of '${snakeName}_bloc.dart';

@freezed
sealed class ${blocName}State with _$${blocName}State {
  const factory ${blocName}State.initial() = ${blocName}Initial;
  const factory ${blocName}State.loading() = ${blocName}Loading;
  const factory ${blocName}State.loaded() = ${blocName}Loaded;
  const factory ${blocName}State.error({required String message}) = ${blocName}Error;
}
`;
}

function main() {
  const argv = process.argv.slice(2);
  if (argv.length < 1) {
    process.stderr.write('Usage: bloc_template.py <BlocName> [bloc|event|state]\n');
    process.stderr.write('Example: bloc_template.py Home\n');
    process.stderr.write('Example: bloc_template.py Home bloc\n');
    process.stderr.write('Example: bloc_template.py Home event\n');
    process.stderr.write('Example: bloc_template.py Home state\n');
    process.exit(1);
  }

  let blocName = argv[0];
  if (blocName.endsWith('Bloc')) {
    blocName = blocName.slice(0, -4);
  }

  const fileType = argv[1] ?? 'bloc';

  if (fileType === 'bloc') {
    console.log(generateBloc(blocName));
  } else if (fileType === 'event') {
    console.log(generateEvent(blocName));
  } else if (fileType === 'state') {
    console.log(generateState(blocName));
  } else {
    process.stderr.write(`Unknown file type: ${fileType}\n`);
    process.exit(1);
  }
}

main();
